src/cpp/session/modules/SessionRAddins.cpp

```diff
@@ -0,0 +1,315 @@
+         fields["Description"],
+         fields["Binding"]));
+}
+
+void registerAddins(const std::string& pkgName, const FilePath& addinPath)
+{
+   static const boost::regex reSeparator("\\n{2,}");
+   
+   std::string contents;
+   Error error = core::readStringFromFile(addinPath, &contents, string_utils::LineEndingPosix);
+   if (error)
+   {
+      LOG_ERROR(error);
+      return;
+   }
+   
+   boost::sregex_token_iterator it(contents.begin(), contents.end(), reSeparator, -1);
+   boost::sregex_token_iterator end;
+   
+   for (; it != end; ++it)
+   {
+      std::map<std::string, std::string> fields = parseAddinDcf(*it);
+      registerAddin(pkgName, fields);
+   }
+}
+
+void discoverAddins(const std::vector<FilePath>& libPaths)
+{
+   BOOST_FOREACH(const FilePath& libPath, libPaths)
+   {
+      if (!libPath.exists())
+         continue;
+      
+      // get the children in this directory
+      std::vector<FilePath> pkgPaths;
+      Error error = libPath.children(&pkgPaths);
+      if (error)
+      {
+         LOG_ERROR(error);
+         continue;
+      }
+      
+      BOOST_FOREACH(const FilePath& pkgPath, pkgPaths)
+      {
+         // construct the addin path and check if it exists
+         FilePath addinPath = pkgPath.complete("rstudio/addins.dcf");
+         if (!addinPath.exists())
+            continue;
+         
+         std::string pkgName = pkgPath.filename();
+         registerAddins(pkgName, addinPath);
+      }
+   }
+}
+
+void onDeferredInit(bool newSession)
+{
+   discoverAddins(getLibPaths());
+}
+
+Error getRAddins(const json::JsonRpcRequest& request,
+                 json::JsonRpcResponse* pResponse)
+{
+   discoverAddins(getLibPaths());
+   pResponse->setResult(addinRegistry().toJson());
+   return Success();
+}
+
+Error noSuchAddin(const ErrorLocation& errorLocation)
+{
+   return systemError(boost::system::errc::invalid_argument, errorLocation);
+}
+
+Error executeRAddin(const json::JsonRpcRequest& request,
+                    json::JsonRpcResponse* pResponse)
+{
+   std::string commandId;
+   Error error;
+   
+   error = json::readParams(request.params, &commandId);
+   if (error)
+   {
+      LOG_ERROR(error);
+      return error;
+   }
+   
+   std::vector<std::string> splat = core::algorithm::split(commandId, "::");
+   if (splat.size() != 2)
+   {
+      LOG_ERROR_MESSAGE("unexpected command id '" + commandId + "'");
+      return Success();
+   }
+   
+   std::string pkgName = splat[0];
+   std::string cmdName = splat[1];
+   
+   if (!addinRegistry().contains(pkgName, cmdName))
+   {
+      LOG_ERROR_MESSAGE("no command with id '" + commandId + "'");
+      pResponse->setError(noSuchAddin(ERROR_LOCATION));
+      return Success();
+   }
+   
+   SEXP fnSEXP = r::sexp::findFunction(cmdName, pkgName);
+   if (fnSEXP == R_UnboundValue)
+   {
+      LOG_ERROR_MESSAGE("no function '" + cmdName + "' found in package '" + pkgName + "'");
+      pResponse->setError(noSuchAddin(ERROR_LOCATION));
```

Comment:
Do you want to setClientInfo in this case so that the client actually knows that the error was no addin found?